<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>SCRuB R Tutorial</title>

<script src="R-tutorial_files/header-attrs-2.20/header-attrs.js"></script>
<script src="R-tutorial_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="R-tutorial_files/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="R-tutorial_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="R-tutorial_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="R-tutorial_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="R-tutorial_files/navigation-1.1/tabsets.js"></script>
<link href="R-tutorial_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="R-tutorial_files/highlightjs-9.12.0/highlight.js"></script>
<link href="R-tutorial_files/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="R-tutorial_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SCRuB</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fab fa-r-project"></span>
     
    Package
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="R-tutorial.html">Tutorial</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="glyphicon glyphicon-zoom-in"></span>
     
    QIIME2 Plugin
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="qiime2-intro.html">Inputs &amp; Outputs</a>
    </li>
    <li>
      <a href="QIIME2-Tutorial.html">Tutorial</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-binoculars"></span>
     
    Related documents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://www.nature.com/articles/s41587-023-01696-w">SCRuB Manuscript</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="contribute.html">
    <span class="fa fa-shower"></span>
     
    Contribute to SCRuB!
  </a>
</li>
<li>
  <a href="https://github.com/Shenhav-and-Korem-labs/SCRuB/issues">
    <span class="fa fa-comment"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/Shenhav-and-Korem-labs/SCRuB/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">SCRuB R Tutorial</h1>

</div>


<style>
    body { background-color: #162138;
            text-color: whitesmoke}
</style>
<p><img src='images/SCRuB_logo.png' align="right" height="139" /></p>
<p>In this tutorial, we demonstrate how SCRuB should be used to remove
contamination from microbial samples. All you need to run SCRuB is a
count matrix of control and non-control samples. To run the SCRuB with
its full spatial component, which is strongly recommended, the well
locations of each sample must also be provided.</p>
<hr />
<div id="getting-ready" class="section level1">
<h1>Getting ready</h1>
<p>We start by loading the necessary packages, such as
<code>SCRUB</code>. We also include the use of <code>tidyverse</code>
within our tutorial</p>
<pre class="r"><code>set.seed(1)
library(SCRuB)</code></pre>
<p>The primary input for SCRuB is a matrix, in which each row represents
a sample, and each column represents each sample’s read counts that
correspond to a certain taxa (e.g ASV, OTU, species). In general, we
recommend to run SCRuB on the most granular version of the data possible
(e.g., ASV, OTU, mOTU), and so data should only be grouped to higher
orders of phylogeny (e.g. genus, family…) <em>after</em> completing the
SCRuB pipeline.</p>
<p>Within this tutorial, we use processed samples publicly available
from <a href="https://qiita.ucsd.edu/">Qiita</a>, from a dataset of
plasma samples, made public by <a
href="https://www.nature.com/articles/s41586-020-2095-1">Poore et
al</a>. In the ensuing code, we demonstrate how SCRuB can be used to
remove contamination from these plasma samples using control samples and
publicly available metadata provided by the authors.</p>
</div>
<div id="section" class="section level1 tabset">
<h1 class="tabset"></h1>
<div id="inputs" class="section level2">
<h2>Inputs</h2>
<p>The required inputs to <em>SCRuB</em> are composed of one count
matrix, and one metadata matrix (or data frame):</p>
<ol style="list-style-type: decimal">
<li><strong><span style="color:lightblue">data</strong> - An (
<code>n_samples</code> + <code>n_controls</code> ) x
<code>n_taxa</code>count matrix, where <code>n_samples</code> is the
number of samples, <code>n_controls</code> is the number of negative
controls, and <code>n_taxa</code> is the number of features in the count
matrix. Row names are the sample ids. Column names are the taxa ids.
Every consecutive column contains read counts for each sample. Note that
this order must be respected.</li>
</ol>
<p>Alternatively, SCRuB supports inputs as 1) a biom object, which
contains the x <code>n_taxa</code> x ( <code>n_samples</code> +
<code>n_controls</code> ) read count matrix; if no metadata input is
specified directly in the function call, <code>SCRuB</code> will check
for a metadata table within the biom object; 2) a <code>character</code>
input specifying the path to a file where the data object is stored;
SCRuB supports <code>.biom</code>, <code>.tsv</code>, and
<code>.csv</code> files.</p>
<p>count matrix (first 4 rows and columns):</p>
<pre class="r"><code>data &lt;- read.csv(&#39;plasma_data.csv&#39;, row.names=1) %&gt;% as.matrix()
data[30:42, 20:24]</code></pre>
<pre><code>##                 G000005845 G000006275 G000006335 G000006355 G000006405
## 12692.150171             0          0          0          0          0
## 12692.150178             1          0          0          0          1
## 12692.150182             0          0          0          1          0
## 12692.150186             0          0          0          0          5
## 12692.150199             0          0          0          0          0
## 12692.150209             0          0          0          0          0
## 12692.150215             0          0          0          6          0
## 12692.150224             0          0          0          0          0
## 12692.150229            45          0          0          0          0
## 12692.150230             0          0          0          0          0
## 12692.150238             1          0          0          0          0
## 12692.150242             3          0          0          0          1
## 12692.Control.1          9          0          0          0          3</code></pre>
<ol start="2" style="list-style-type: decimal">
<li><strong><span style="color:lightblue">metadata</strong> - (n_samples
+ n_controls x 2, or (n_samples + n_controls x 3. The columns must be
ordered as described below, with the third column being optional. The
three columns are 1) a boolean indicator identifying which entries
correspond to control samples; 2) a string column denoting the types of
control samples; and (optionally) 3) the plate location of each
sample.</li>
</ol>
<p>Alternatively, SCRuB will accept a <code>character</code> path to a
<code>.csv</code> or <code>.tsv</code> file containing the metadata.</p>
<pre class="r"><code>metadata &lt;- read.csv(&#39;plasma_metadata.csv&#39;, row.names=1)
metadata[30:42,]</code></pre>
<pre><code>##                 is_control                  sample_type sample_well
## 12692.150171         FALSE                       plasma         F22
## 12692.150178         FALSE                       plasma         H22
## 12692.150182         FALSE                       plasma         P16
## 12692.150186         FALSE                       plasma         J22
## 12692.150199         FALSE                       plasma         H18
## 12692.150209         FALSE                       plasma         L22
## 12692.150215         FALSE                       plasma         P22
## 12692.150224         FALSE                       plasma         D14
## 12692.150229         FALSE                       plasma         N22
## 12692.150230         FALSE                       plasma         F14
## 12692.150238         FALSE                       plasma         B24
## 12692.150242         FALSE                       plasma         D24
## 12692.Control.1       TRUE control blank DNA extraction          A3</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>(optional)
<strong><span style="color:lightblue">control_order</strong> - vector,
default NA. If specified, outlines the order of controls to run
decontamination in. Input as a vector, of which each element must also
be found in the metadata’s second column. If not specified, all control
types found in <code>metadata</code> will be run sequentially based on
their order from that table. In thise example, our dataset includes the
following types of controls:</li>
</ol>
<pre class="r"><code>metadata %&gt;% filter(is_control) %&gt;% pull(sample_type) %&gt;% unique()</code></pre>
<pre><code>## [1] &quot;control blank DNA extraction&quot; &quot;control blank library prep&quot;</code></pre>
<p>Based on the analysis from SCRuB’s manuscript, we recommend to
perform decontamination in the order in which contaminants are
introduced. In this case, this corresponds to the following control
order:</p>
<pre class="r"><code>control_order &lt;- c(&quot;control blank DNA extraction&quot;, &quot;control blank library prep&quot;)</code></pre>
</div>
<div id="running-scrub" class="section level2">
<h2>Running SCRuB</h2>
<p>To run SCRuB, we only need to execute a single function:</p>
<pre class="r"><code>scr_out &lt;- SCRuB( data[,1:750], # temporarily shortening feature space to shorten tutorial runtime as the documentation is being developped...
                  metadata, 
                  control_order = control_order
                  )</code></pre>
<pre><code>## [1] &quot;Incorporating the well metadata to track well-to-well leakage!&quot;
## [1] &quot;SCRuBbing away contamination in the control blank DNA extraction controls...&quot;
## [1] &quot;SCRuBbing away contamination in the control blank library prep controls...&quot;</code></pre>
<p>Alternatively, SCRuB can be run using <code>biom</code> objects, and
by inputting the paths to the input files. The file formats for the
inputs can be found <a
href="https://github.com/Shenhav-and-Korem-labs/SCRuB/tree/gh-pages_tmp">here</a>.</p>
<pre class="r"><code>scr_out &lt;- SCRuB( &#39;plasma-data.biom&#39;,
                  &#39;plasma_metadata.csv&#39;, 
                  control_order =  c(&#39;control blank DNA extraction&#39;, &#39;control blank library prep&#39;)
                  )</code></pre>
<pre><code>## [1] &quot;Incorporating the well metadata to track well-to-well leakage!&quot;
## [1] &quot;SCRuBbing away contamination in the control blank DNA extraction controls...&quot;
## [1] &quot;SCRuBbing away contamination in the control blank library prep controls...&quot;</code></pre>
</div>
<div id="outputs" class="section level2">
<h2>Outputs</h2>
<p>SCRuB outputs a list containing the following components:</p>
<ol style="list-style-type: decimal">
<li><strong><span style="color:lightblue">decontaminated_samples</strong>
- an n_samples x n_taxa count matrix, representing the decontaminated
taxonomic compositions. In most cases, this is what you are interested
in for downstream analyses.</li>
</ol>
<pre class="r"><code>decontaminated_samples &lt;- scr_out$decontaminated_samples
decontaminated_samples[30:41, 20:24]</code></pre>
<pre><code>##              G000005845 G000006275 G000006335 G000006355 G000006405
## 12692.150171          0          0          0          0          0
## 12692.150178          0          0          0          0          0
## 12692.150182          0          0          0          1          0
## 12692.150186          0          0          0          0          0
## 12692.150199          0          0          0          0          0
## 12692.150209          0          0          0          0          0
## 12692.150215          0          0          0          6          0
## 12692.150224          0          0          0          0          0
## 12692.150229          0          0          0          0          0
## 12692.150230          0          0          0          0          0
## 12692.150238          0          0          0          0          0
## 12692.150242          0          0          0          0          0</code></pre>
<ol start="2" style="list-style-type: decimal">
<li><strong><span style="color:lightblue">p</strong> - an n_samples
vector representing the estimate proportion of each observed sample that
was not contamination, as described in SCRuB’s methods. A dataset that
had no contamination would have a<code>p</code> of 1s, while a dataset
of entirely contamination would have a <code>p</code> of 0.</li>
</ol>
<pre class="r"><code>(1 - scr_out$p) %&gt;% 
  boxplot(ylab=&#39;Level of contamination&#39;)</code></pre>
<p><img src="R-tutorial_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<ol start="3" style="list-style-type: decimal">
<li><strong><span style="color:lightblue">inner_iterations</strong> -
results from SCRuB’s intermediary steps, which includes one list entry
per type of control. Each iteration contains:</li>
</ol>
<p>(3.1)
<strong><span style="color:lightblue">decontaminated_samples</strong> -
a n_samples x n_taxa count matrix, representing the taxonomic
composition of the samples after the specified round of
decontamination.</p>
<pre class="r"><code>scr_out$inner_iterations$`control blank DNA extraction`$decontaminated_samples[30:41,20:24]</code></pre>
<pre><code>##              G000005845 G000006275 G000006335 G000006355 G000006405
## 12692.150171          0          0          0          0          0
## 12692.150178          0          0          0          0          1
## 12692.150182          0          0          0          1          0
## 12692.150186          0          0          0          0          5
## 12692.150199          0          0          0          0          0
## 12692.150209          0          0          0          0          0
## 12692.150215          0          0          0          6          0
## 12692.150224          0          0          0          0          0
## 12692.150229         47          0          0          0          0
## 12692.150230          0          0          0          0          0
## 12692.150238          0          0          0          0          0
## 12692.150242          0          0          0          0          0</code></pre>
<p>(3.2) <strong><span style="color:lightblue">p</strong> – an n_sample
vector representing the estimate proportion of each observed sample that
was not contamination, based on the specified type of process
control</p>
<pre class="r"><code>(1 - scr_out$inner_iterations$`control blank DNA extraction`$p ) %&gt;%
  boxplot(ylab=&#39;Level of contamination&#39;)</code></pre>
<p><img src="R-tutorial_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>(3.3) <strong><span style="color:lightblue">alpha</strong> - an
n_control x ( n_sample + 1 ) matrix, representing the estimated
contribution of the contaminant and each sample to each control, where
the (n_sample + 1)th column represents the contribution from the
contamination to the control. Each row of alpha sums to 1, with each
entry of the (n_sample + 1)th column being 1 means there is zero
estimated well leakage, while entries close to zero would indicate there
is a high level of well leakage. It is strongly recommended to
incorporate the well metadata of each sample into SCRuB, as this makes
it possible ot directly account for potential well leakage into negative
controls, reducing the risk of wrongly removing certain species during
decontamination.</p>
<p>In this case, <code>SCRuB</code> estimated a very high level of
leakage for many DNA extraction samples, with a median estimated
well-to-well leakage of <code>60%</code>.</p>
<pre class="r"><code>boxplot( 1 - scr_out$inner_iterations$`control blank DNA extraction`$alpha[, ncol(scr_out$inner_iterations$`control blank DNA extraction`$alpha)], 
         ylab=&#39;Level of well leakage in process controls&#39;)</code></pre>
<p><img src="R-tutorial_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>(3.4) <strong><span style="color:lightblue">gamma</strong> an n_taxa
vector representing the estimated relative abundance of the
contamination community, as described in SCRuB’s methods.</p>
<pre class="r"><code>scr_out$inner_iterations$`control blank DNA extraction`$gamma %&gt;% 
  plot(ylab=&#39;Relative abundance&#39;, xlab=&#39;Different OTUs in contamination source&#39;)</code></pre>
<p><img src="R-tutorial_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>(3.5) <strong><span style="color:lightblue">loglikelihood</strong> -
float. The layer-specific log-likelihood of the inputted dataset based
on SCRuB’s fitted parameters.</p>
<pre class="r"><code>scr_out$inner_iterations$`control blank DNA extraction`$loglikelihood</code></pre>
<pre><code>## [1] -899894985</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
